src/test/java/com/microsoft/sqlserver/jdbc/connection/NativeMSSQLDataSourceTest.java

```diff
@@ -66,6 +84,54 @@ public void testDSNormal() throws ClassNotFoundException, IOException, SQLExcept
         try (Connection conn = ds.getConnection()) {}
     }
 
+    @Tag(Constants.xSQLv11)
+    @Tag(Constants.xSQLv12)
+    @Tag(Constants.xSQLv14)
+    @Tag(Constants.xSQLv15)
+    @Tag(Constants.xAzureSQLDW)
+    @Tag(Constants.reqExternalSetup)
+    @Test
+    public void testDSPooledConnectionAccessTokenCallback() throws SQLException {
+        SQLServerAccessTokenCallback callback = new SQLServerAccessTokenCallback() {
+            @Override
+            public SqlAuthenticationToken getAccessToken(String spn, String stsurl) {
+                String scope = spn + "/.default";
+                Set<String> scopes = new HashSet<>();
+                scopes.add(scope);
+
+                try {
+                    ExecutorService executorService = Executors.newSingleThreadExecutor();
+                    IClientCredential credential = ClientCredentialFactory.createFromSecret(accessTokenSecret);
+                    ConfidentialClientApplication clientApplication = ConfidentialClientApplication
+                            .builder(accessTokenClientId, credential).executorService(executorService)
+                            .authority(stsurl).build();
+                    CompletableFuture<IAuthenticationResult> future = clientApplication
+                            .acquireToken(ClientCredentialParameters.builder(scopes).build());
+
+                    IAuthenticationResult authenticationResult = future.get();
+                    String accessToken = authenticationResult.accessToken();
+                    long expiresOn = authenticationResult.expiresOnDate().getTime();
+
+                    return new SqlAuthenticationToken(accessToken, expiresOn);
+
+                } catch (Exception e) {
+                    fail(TestResource.getResource("R_unexpectedException") + e.getMessage());
+                }
+                return null;
+            }
+        };
+
+        SQLServerConnectionPoolDataSource ds = new SQLServerConnectionPoolDataSource();
+        AbstractTest.updateDataSource(connectionString, ds);
+        ds.setUser("");
+        ds.setPassword("");
+        ds.setAccessTokenCallback(callback);
+
+        // Callback should provide valid token on connection open for all new connections
+        try (Connection conn1 = (SQLServerConnection) ds.getConnection()) {}
+        try (Connection conn2 = (SQLServerConnection) ds.getConnection()) {}
```

Comment:
There are a few more validations you can do here.

1. When the access token hasn't expired, validate that the underlying connection is the same between conn1 and conn2.
2. When the access token has expired (manually set the date to something that will trigger checkIfNeedNewAccessToken in Util.java), validate that the underlying connection is different between conn1 and conn2.
3. Connect, run a query, kill the connection and run another query. ICR should reconnect and get an access token using the callback.